File: net-event-bus/examples/registry.py
"""The service registry you no longer run.

Four in-process mesh nodes over loopback UDP: two providers announce the same
capability, a caller discovers them and ranks them locally, a third provider
appears, and the caller's next lookup ranks it first — with no registry process,
no health-check poller, no config reload and no announcement to any address.

Run:

    python registry.py

Expected final line: ``RESULT ok providers=3 joined=1 best_moved=1``
"""

from __future__ import annotations

import threading
import time
from collections.abc import Callable

from net import NetMesh

# 64 hex characters = 32 bytes. Every node in a mesh shares it.
PSK = "42" * 32

CONVERGE_S = 5.0


def build(seed: int) -> NetMesh:
    return NetMesh(
        "127.0.0.1:0",
        PSK,
        identity_seed=bytes([seed]) * 32,
        # Tight heartbeat, like the binding's own tests, so gossip settles fast.
        heartbeat_interval_ms=200,
    )


def handshake(responder: NetMesh, initiator: NetMesh) -> None:
    """One side connects, the other accepts.

    Both calls block, so the responder's ``accept`` runs on its own thread —
    the same shape the binding's own test fixture uses. ``accept`` must be
    registered before ``start``.
    """
    errors: list[BaseException] = []

    def accept() -> None:
        try:
            responder.accept(initiator.node_id)
        except BaseException as error:  # noqa: BLE001 - surfaced below
            errors.append(error)

    thread = threading.Thread(target=accept, daemon=True)
    thread.start()
    time.sleep(0.05)
    initiator.connect(responder.local_addr, responder.public_key, responder.node_id)
    thread.join(timeout=5.0)
    if thread.is_alive():
        raise RuntimeError("handshake timed out")
    if errors:
        raise errors[0]


def until(what: str, predicate: Callable[[], bool]) -> None:
    """Poll the caller's local capability fold until ``predicate`` holds."""
    deadline = time.monotonic() + CONVERGE_S
    while time.monotonic() < deadline:
        if predicate():
            return
        time.sleep(0.025)
    raise TimeoutError(f"timed out waiting for {what}")


def main() -> None:
    # Announcements reach directly-connected peers, so the caller connects to
    # every provider it wants to see. There is no directory to join.
    a = build(0xA1)  # provider, 16 GB
    b = build(0xB2)  # provider, 64 GB
    c = build(0xC3)  # provider that appears later, 256 GB
    caller = build(0xD4)

    # Every accept() completes before any start().
    handshake(b, a)  # A <-> B
    handshake(a, caller)  # A <-> caller
    handshake(caller, b)  # B <-> caller
    handshake(caller, c)  # C <-> caller

    a.start()
    b.start()
    c.start()
    caller.start()

    try:
        # The two original providers announce. That is the entire registration.
        a.announce_capabilities({"hardware": {"memory_gb": 16}, "tags": ["api"]})
        b.announce_capabilities({"hardware": {"memory_gb": 64}, "tags": ["api"]})

        query = {"require_tags": ["api"]}
        until("both providers to appear", lambda: len(caller.find_nodes(query)) == 2)
        providers = len(caller.find_nodes(query))

        # Ranking is local and free: prefer more memory, so the bigger machine
        # wins without any scheduler deciding it centrally.
        requirement = {"filter": query, "prefer_more_memory": 1.0}
        first = caller.find_best_node(requirement)
        if first is None:
            raise RuntimeError("no provider matched")

        # A new provider appears: it announces once and is immediately
        # addressable — no registry entry, no discovery config, no reload.
        c.announce_capabilities({"hardware": {"memory_gb": 256}, "tags": ["api"]})
        until("the new provider to appear", lambda: len(caller.find_nodes(query)) == 3)
        after = len(caller.find_nodes(query))

        best = caller.find_best_node(requirement)
        if best is None:
            raise RuntimeError("no provider matched after the join")

        moved = 1 if first != best else 0
        print(f"providers found at first lookup: {providers}")
        print(f"ranked best:                     0x{first:x}")
        print(f"providers after the join:        {after}")
        print(f"ranked best:                     0x{best:x}")

        print(f"RESULT ok providers={after} joined=1 best_moved={moved}")
    finally:
        a.shutdown()
        b.shutdown()
        c.shutdown()
        caller.shutdown()


if __name__ == "__main__":
    main()

File: net-event-bus/examples/registry.ts
//! The service registry you no longer run.
//!
//! Four in-process mesh nodes over loopback UDP: two providers announce the
//! same capability, a caller discovers them and ranks them locally, a third
//! provider appears, and the caller's next lookup ranks it first — with no
//! registry process, no health-check poller, no config reload and no
//! announcement to any address.
//!
//! Run (from `net/crates/net/sdk-ts`, where `tsconfig.skill-example.json`
//! includes it):
//!
//!   npx tsc --noEmit -p tsconfig.skill-example.json
//!
//! Expected final line: `RESULT ok providers=3 joined=1 best_moved=1`

import { MeshNode } from '@net-mesh/sdk';

/** 64 hex characters = 32 bytes. Every node in a mesh shares it. */
const PSK = '42'.repeat(32);

const CONVERGE_MS = 5_000;

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

async function build(seed: number): Promise<MeshNode> {
  return MeshNode.create({
    bindAddr: '127.0.0.1:0',
    psk: PSK,
    identitySeed: Buffer.alloc(32, seed),
    // Tight heartbeat, like the SDK's own tests, so gossip settles quickly.
    heartbeatIntervalMs: 200,
  });
}

/**
 * One side connects, the other accepts. `accept` MUST be registered before
 * `start`; both promises resolve only after the handshake completes, so
 * awaiting the pair is the wait-until-connected primitive.
 */
async function handshake(responder: MeshNode, initiator: MeshNode): Promise<void> {
  const addr = responder.localAddr();
  const pub = responder.publicKey();
  const responderId = responder.nodeId();
  await Promise.all([
    responder.accept(initiator.nodeId()),
    (async () => {
      await sleep(50);
      await initiator.connect(addr, pub, responderId);
    })(),
  ]);
}

/** Poll the caller's local capability fold until `pred` holds, or give up. */
async function until(what: string, pred: () => boolean): Promise<void> {
  const deadline = Date.now() + CONVERGE_MS;
  while (Date.now() < deadline) {
    if (pred()) return;
    await sleep(25);
  }
  throw new Error(`timed out waiting for ${what}`);
}

async function main(): Promise<void> {
  // Announcements reach directly-connected peers, so the caller connects to
  // every provider it wants to see. There is no directory to join.
  const a = await build(0xa1); // provider, 16 GB
  const b = await build(0xb2); // provider, 64 GB
  const c = await build(0xc3); // provider that appears later, 256 GB
  const caller = await build(0xd4);

  // Every accept() completes before any start().
  await handshake(b, a); // A <-> B
  await handshake(a, caller); // A <-> caller
  await handshake(caller, b); // B <-> caller
  await handshake(caller, c); // C <-> caller

  await a.start();
  await b.start();
  await c.start();
  await caller.start();

  // The two original providers announce. That is the entire registration.
  await a.announceCapabilities({ hardware: { memoryGb: 16 }, tags: ['api'] });
  await b.announceCapabilities({ hardware: { memoryGb: 64 }, tags: ['api'] });

  const filter = { requireTags: ['api'] };
  await until('both providers to appear', () => caller.findNodes(filter).length === 2);
  const providers = caller.findNodes(filter).length;

  // Ranking is local and free. Prefer more memory, so the bigger machine wins
  // without any scheduler deciding it centrally. `findBestNode` returns null
  // for "no match" — 0n is a real node id.
  const req = { filter, preferMoreMemory: 1 };
  const first = caller.findBestNode(req);
  if (first === null) throw new Error('no provider matched');

  // A new provider appears: it announces once and is immediately addressable.
  await c.announceCapabilities({ hardware: { memoryGb: 256 }, tags: ['api'] });
  await until('the new provider to appear', () => caller.findNodes(filter).length === 3);
  const after = caller.findNodes(filter).length;

  const best = caller.findBestNode(req);
  if (best === null) throw new Error('no provider matched after the join');

  const moved = first !== best ? 1 : 0;
  console.log(`providers found at first lookup: ${providers}`);
  console.log(`ranked best:                     0x${first.toString(16)}`);
  console.log(`providers after the join:        ${after}`);
  console.log(`ranked best:                     0x${best.toString(16)}`);

  console.log(`RESULT ok providers=${after} joined=1 best_moved=${moved}`);

  await a.shutdown();
  await b.shutdown();
  await c.shutdown();
  await caller.shutdown();
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
